README.md

```diff
@@ -1,2 +1,159 @@
 # linux-prewarm-process-rfc
 WIP - For now it's just an idea of how to launch processes ahead of time for bash scripts
+
+# context / problem
+shell scripts (in any flavour/variant) are incredibly composable due to unix pipelines. 
+This led many people to develop a plethora CLI utilies for all sorts of problems out there
+(curl, grep, sed, cat - and the list goes on and on).
+
+The problem is that there is a long list of housekeeping actions that the operating system
+must perform before creating processes such as:
+
+- dynamic linker kicks in and load dependent libraries
+- allocated physical space to accomodate virtual addresses
+- read executable from disk
+- allocate pid
+- put process on OS scheduler (usually CFS)
+- just to name a few.... there is actually much more than this.
+
+Also, it's worth mentioning that while killing processes a lot of the aforementioned
+tasks must be undone by the OS which also takes time.
+
+As a result, even small shell script files tend to be noticible slower than equivalent
+programs written in other programming languages. 
+
+For instance, imagine that you are writing a AWS lambda runtime in Bash which doesn't do
+anything else other than echoing back whatever was sent through the input payload:
```

Comment:
Is there an example we could add here that would be testable for the average person (aka me) without needing AWS credentials? E.g., the same idea as below, but a much simpler thing (e.g, pinging a jokes api to get a joke).

Reply:
Good point. There reason I used this one is because I had just given a presentation on the topic (so I knew where the source code was lol). I put together two as-equivalent-as-possible programs but one is written in bash and the other is the "binary-monolith" thing in C. 

I also added the benchmark on executing each of them... that already gives us indications of things that we can solve such as reducing number of page faults but also things that we can't solve such as number of branch-misses. (I'm probably getting ahead of myself now... we can discuss it at length once we are at the development discussion phase).

## Bash

```bash
#!/bin/bash

echo "test" > /tmp/file.txt
cat /tmp/file.txt
rm -f /tmp/file.txt
```

**Performance:**
```
echo 0 > /proc/sys/kernel/nmi_watchdog
perf stat -r 100 -B ./test-script.sh
echo 1 > /proc/sys/kernel/nmi_watchdog

 Performance counter stats for './test-script.sh' (100 runs):

              2.05 msec task-clock                       #    0.755 CPUs utilized            ( +-  2.96% )
                 2      context-switches                 #  839.719 /sec                     ( +-  2.32% )
                 0      cpu-migrations                   #    0.000 /sec                   
               367      page-faults                      #  154.089 K/sec                    ( +-  0.08% )
         6,721,109      cycles                           #    2.822 GHz                      ( +-  1.31% )
           916,703      stalled-cycles-frontend          #   12.58% frontend cycles idle     ( +-  2.95% )
           919,249      stalled-cycles-backend           #   12.61% backend cycles idle      ( +-  2.79% )
         6,530,299      instructions                     #    0.90  insn per cycle         
                                                  #    0.16  stalled cycles per insn  ( +-  0.15% )
         1,581,210      branches                         #  663.886 M/sec                    ( +-  0.13% )
            38,750      branch-misses                    #    2.44% of all branches          ( +-  0.31% )

         0.0027217 +- 0.0000720 seconds time elapsed  ( +-  2.64% )
```

## C

```C
#include <stdio.h>
#include <stdlib.h>

#define FILENAME        "/tmp/file.txt"

int main(void) {    
        FILE *fp;
        char buff[255];

        // echo "test" > /tmp/file.txt
        fp = fopen(FILENAME, "w+");
        fputs("paulo\n", fp);
        fclose(fp);

        // cat /tmp/file.txt
        fp = fopen(FILENAME, "r");
        fscanf(fp, "%s", buff);
        printf("%s\n", buff );
        fclose(fp);

        // rm -f /tmp/file.txt
        remove(FILENAME);

        return 0;
}
```

**Performance:**
```
echo 0 > /proc/sys/kernel/nmi_watchdog
perf stat -r 100 -B ./test-script-c
echo 1 > /proc/sys/kernel/nmi_watchdog

 Performance counter stats for './test-script-c' (100 runs):

              0.31 msec task-clock                       #    0.486 CPUs utilized            ( +-  2.23% )
                 0      context-switches                 #    0.000 /sec                   
                 0      cpu-migrations                   #    0.000 /sec                   
                60      page-faults                      #  159.608 K/sec                    ( +-  0.15% )
           884,584      cycles                           #    2.353 GHz                      ( +-  1.61% )
            55,538      stalled-cycles-frontend          #    5.73% frontend cycles idle     ( +-  6.72% )
            95,613      stalled-cycles-backend           #    9.87% backend cycles idle      ( +-  3.20% )
           947,810      instructions                     #    0.98  insn per cycle         
                                                  #    0.12  stalled cycles per insn  ( +-  0.15% )
           235,672      branches                         #  626.919 M/sec                    ( +-  0.14% )
             5,603      branch-misses                    #    2.41% of all branches          ( +-  0.79% )

         0.0006380 +- 0.0000139 seconds time elapsed  ( +-  2.18% )
```